File: Desktop/Code/NewFolder/5.1/list.cpp
#include "list.h"
#include <iostream>

using namespace std;

void deleteList(List *list) 
{
    ListElement *element = list->head;
    while (element->next != nullptr) { 
        ListElement *temp = element->next->next;
        delete element->next;
        element->next = temp;
    }
    delete list->head;
}

void insert(List *list, int value) 
{
    ListElement *element = list->head;
    if (list->head == nullptr) {
        ListElement *newElement = new ListElement;
        newElement->next = nullptr;
        newElement->value = value;
        list->head = newElement;
    }
    else {
        while (element->next != nullptr) {
            if (value  < element->value && value > element->next->value) {
                ListElement *temp = element->next;
                ListElement *newElement = new ListElement;
                newElement->next = nullptr;
                newElement->value = value;
                element->next = newElement;
                newElement->next = temp;
            }
            else {
                element = element->next;
            }
        }
    }
}

void print(List *list, int value) {
    ListElement *element = list->head;
    while (element->next != nullptr) {
        cout << element->value;
        element = element->next;
    }
}

int deleteElement(List *list, int value) 
{
    ListElement *element = list->head;
    if(list->head->value == value) {
        cout << "НЕЛЬЗЯ УДАЛЯТЬ ЭЛЕМЕНТ ИЗ ГОЛОВЫ СПИСКА ЭТО ОШИБКА";
        return 0;
    }
    while (element->next != nullptr) {
        if (element->next->value == value) {
            ListElement *temp = element->next->next;
            delete element->next;
            element->next = temp;
            return 0;
        }
        else {
            element = element->next; 
        }
    }
    cout << "В списке нет такого элемента";
}

File: 2Semester/2homework/2.2/List.cs
using System;

namespace _2._2
{
    public class List
    {
        private class ListElement
        {
            public int Value
            {
                get;
                set;
            }
            public ListElement Next
            {
                get;
                set;
            }
        }

        private ListElement head = null;

        public void Print()
        {
            Console.WriteLine("Список: ");
            var currentElement = head;
            while (currentElement.Next != null)
            {
                Console.WriteLine(currentElement.Value + " ");
                currentElement = currentElement.Next;
            }
            Console.WriteLine(currentElement.Value + " ");
        }

        public void Add(int value)
        {
            var newElement = new ListElement()
            {
                Value = value,
                Next = null
            };

            if (head == null)
            {
                head = newElement;
                return;
            }

            var currentElement = head;
            while (currentElement.Next != null)
            {
                currentElement = currentElement.Next;
            }
            currentElement.Next = newElement;
        }

        public bool BinarySearch(int value)
        {
            var currentElement = head;
            while (currentElement.Next != null)
            {
                if (currentElement.Value == value)
                {
                    return true;
                }
                currentElement = currentElement.Next;
            }
            return false;
        }

        public void Remove(int value)
        {
            if (head.Value == value)
            {
                head = head.Next;
                return;
            }
            var currentElement = head;
            while (currentElement.Next != null)
            {
                if (currentElement.Next.Value == value)
                {
                    if (currentElement.Next.Next == null)
                    {
                        currentElement.Next = null;
                        return;
                    }
                    var dop = currentElement.Next.Next;
                    currentElement.Next = null;
                    currentElement.Next = dop;
                    return;
                }
                currentElement = currentElement.Next;
            }
            Console.WriteLine("Нельзя удалить элемент, которого нет :)))");
        }
    }
}
